### ISSUE TITLE
Possible problem of concurrence in DefaultTokenServices

### ISSUE THREAD

**Issue by ismaelgomescosta** [spring-projects/spring-security-oauth#502]
In the following code:

``` java

OAuth2AccessToken existingAccessToken = tokenStore.getAccessToken(authentication);
        OAuth2RefreshToken refreshToken = null;
        if (existingAccessToken != null) {
            if (existingAccessToken.isExpired()) {
                if (existingAccessToken.getRefreshToken() != null) {
                    refreshToken = existingAccessToken.getRefreshToken();
                    // The token store could remove the refresh token when the
                    // access token is removed, but we want to
                    // be sure...
                    tokenStore.removeRefreshToken(refreshToken);
                }
                tokenStore.removeAccessToken(existingAccessToken);
            }
            else {
                // Re-store the access token in case the authentication has changed
                tokenStore.storeAccessToken(existingAccessToken, authentication);
                return existingAccessToken;
            }
        }
```

I'm using the default implementation, but I have a problem of concurrence in this case:

Application A = OAuth2 server
Application B = Consumes the Application A and has his own services.
Application C = Consumes the Application B.
Application D = Consumes the Application B.

The user logs in the Application C, and logs in Application D. Both applications (C and D) receive the same token, because the Application B logs in Application A, and the Application A gives the same token to Application B.

So, when the Application C call the refresh token, the Application C receives a new token, but the Application D keeps the first access token, then the user logs out from Application D.

It would be interesting to have a boolean to deactivate the reuse of the access token or simply remove the reuse of access token.

This is my case, I do not know if is a common case or a specific case. If it is a specif case, I can implement my own TokenServices.

Thank you.

**Comment by dsyer** [comment]
If all "Applications" are OAuth2 clients they can't get the same token (in the default `TokenStore` implementations). I'm not sure I follow. Where do you see an access token being re-used across clients? Which `TokenStore` are you using?

**Comment by mundodojava** [comment]
1 - Where do you see an access token being re-used across clients? 

No. I think you misunderstood what I said before because I couldn't explain properly. Let me try again.

I am using ```DefaultTokenServices```.

The Application C and D are not OAuth2 clients.

If you replace Application C and D for one mobile app, the same thing will occur.

If the user logs in two or more cellphones (all mobile app will be the same clientId because are different installations). So, when one of the cellphones calls the refresh token, the user will log out in the other cellphones, and this isn't a good experience for the user.

**Comment by mundodojava** [comment]
I suggest, if possible, add the access token reuse control per client.

**Comment by dsyer** [comment]
If C and D are not OAuth2 clients what are they doing with tokens?

**Comment by mundodojava** [comment]
C and D don't access the tokens directly. C and D access B. B has tokens. However, C and D call a login service of B, and B redirect this login service adding clientId and clientSecret to the request and redirect to A the login. So, when the B returns the token to C and D both receive the same token. 

But this isn't the point. How I said before, this problem could occur with mobile applications or browsers. If the user logs in two browsers, and each browser has a different session, one browser can log out the other one.

**Comment by dsyer** [comment]
> If  the user logs in two browsers, and each browser has a different session, one browser can log out the other one.

Only if B makes the mistake of using an access token as a session identifier. C and D should have their own, independent cookies if they need independent sessions. I'm not sure what the connection is to OAuth yet.

**Comment by mundodojava** [comment]
Okay, C and D have independent cookies and independent sessions, and that's the point.
When the user logs in C and D, B associates the access token retrievied for each session (but it is the same token). 

B needs to associate the access token with the session because B will need to use the access token to call some services of A (the OAuth Server). Each time C or D make requests to B, B will verify if the access token associated to the session is expired. If the token is expired, B will call the refresh token service to renew the session. 

As C and D have the same token, if C calls B before D, and B renew the  token of C, D logs out.

One solution for this problem is: B finds the sessions who are using the access token and change all sessions when B renew the token. But it doesn't sounds good. 

I think each time the user logs in, a new access token should be generated. I don't know if I right understood  the concept of the token of OAuth2. I understand that the access token can be used to call services after the log in, each log in is a unique log in, so each login should have a unique token.

I hope I was more clear this time. This is a little complicated to explain.

If you want to, we can discuss this and make a meeting via Skype.

**Comment by dsyer** [comment]
If the user logs out of B from either of his sessions, C or D, they do not affect each other. So I think your "logout" scenario is a little hazy, but I think I get what you mean now, because  if the token expires that's a different question. You are right that B holds 2 copies of the same access token, and that one of them will refresh before the other, at which point there might be a problem because they won't agree as to the correct access token value, and one will keep cancelling the other by using the refresh token. I haven't actually seen this happen but I can believe it (I'd like someone to confirm it if I don't have time to). 

To be clear, as far as I can see the app will not be broken and the user might not even notice, but there will be a lot more chatter between the app and the auth server than is strictly necessary, Is that what you are worried about? On a positive note, I might have finally understood what the value of `ClientTokenServices` are. (If you use one of those then it will always have the single authoritative value for the current access token for this client.)

I would also like some help understanding how this scenario can arise in practice. C and D might be different browser processes I guess, but why is the user logged on twice to the same app? Different devices I suppose, is the obvious answer. I would still expect it's quite uncommon because mobile devices usually prefer password grant, hence the token will be different and probably should include information about the device in the token, making it unique. But I digress.

**Comment by kulwinders** [comment]
I am also using default implementation of DefaultTokenServices.java in my project. I have question similar to the one mentioned in this thread.
If user is logged in to the application form more than two different browsers using same account info (eg. username and client_id is same in all the cases), then will the access token generated in every cases (on each browser) be same or different.
In my case, If I logged into application using three different browsers(A, B, C) then for two browsers it generates different access token but for the third browser(Say C) it returns the same access token generated already either for A or B.
After debugging I found authenticationKeyGenerator.extractKey(authentication) stement is used inside getAccessToken(authentication) method and returns same key third time which matches the key generated for A or B. So it returns the same existing access token.
As per my requirement,  different access token should be generated, each time user logged into the application and after logout access token is deleted from the db.
My question is why extractKey(authentication) method returns different key in the case of two different browsers using the same account info (username, client_id) and when third comes into picture it returns the key generated previously in either case.

Thanks

**Comment by dsyer** [comment]
AFAIK the default id generator considers the username, client id and scopes. Maybe your third token was for a different scope?

**Comment by kulwinders** [comment]
No, In all the cases username, client id and scopes are same ?

**Comment by mundodojava** [comment]
I don't have time this week to build an example, but as soon as possible I will put a project in my repositories to show such scenario.

Thank you for your attention.

**Comment by kulwinders** [comment]
Ok Thanks ..

**Comment by fsforna** [comment]
Dear All,
the case reported by mundodojava is right. We've noticed the same behaviour. The race condition affects the update inside the storeAccessToken(OAuth2AccessToken token, OAuth2Authentication authentication) method, line 148, class org.springframework.security.oauth2.provider.token.store.JdbcTokenStore.

We reckon that you should change the storeAccessToken method to introduce a check if the access_token already exists. At the moment if the race condition happens a org.springframework.dao.DuplicateKeyException will be raised and this exception is not managed. The execution is in transaction but you've set a optimistic lock, so the update can be executed by other threads.

Regards,
Francesco.

**Comment by dsyer** [comment]
@fsforna that's a different problem. @mundodojava does not have a `DuplicateKeyException` in his scenario. The duplication is on the client (two different clients in fact), and the problem is that they keep cancelling one another's tokens once the first one (duplicate) expires.

I think you just need a transactional token services (should) be automatic since 2.0.7 as long as you `@EnableTransactionManagement`. See issue #276 for more detailed discussion.

**Comment by fsforna** [comment]
We encountered the same issue reported here:

https://github.com/jhipster/generator-jhipster/issues/1759

Do you reckon that if we use the 2.0.8 snapshot-release the issue will be resolved?

Thank you.

**Comment by fsforna** [comment]
Dear @dsyer,
we noticed the @Transactional annotation was moved over methods respect your last commit; by the way, the issue #276 seems to be still open, such as I look to explain in my post.

**Comment by dsyer** [comment]
Issue #276 is closed. Would it make sense for you to try a snapshot if 2.0.7 isn't working?

**Comment by collinpeters** [comment]
This is still an issue as far as I can tell. The way I see it is that you don't need any special setup whatsoever... I just have a `authentication_id` as the primary key (seems to be recommended from schema referenced in tests and tickets on this topic) and a bash script such as this which repeatedly makes the same call to get a token (notice the `&` so there are multiple calls coming in at the same time):
```
#!/bin/bash
for i in {1..10}
do
	echo $i;
	curl -k -X POST -H "Authorization: Basic REDACTED=" "https://localhost:8080/oauth/token?grant_type=password&username=manager&password=password" &
done
```
This results in multiple threads ultimately coming into `JdbcTokenStore.storeAccessToken` at the same time and (despite the `removeAccessToken`) results in multiple `INSERT` calls for the same `authentication_id`. The database constraints and the `@Transactional` don't help here... the database just enforces the rules and a `DuplicateKeyException` is thrown.

The only easy solution I have seen work is to make `DefaultTokenServices.createAccessToken` synchronized. The other solutions of somehow elegantly handling the `DuplicateKeyException` would probably also work. I'm hestitant to use `synchronized` for performance reasons but its the quickest solution for me. Of course, it doesn't help if you have multiple auth servers on one database either. At that point a more elegant solution would be needed with full database locks.
```
public class LockedTokenServices extends DefaultTokenServices {

	@Override
	public synchronized OAuth2AccessToken createAccessToken(OAuth2Authentication authentication)
			throws AuthenticationException {
		return super.createAccessToken(authentication);
	}
}
```
Probably would also need to do the same with `refreshAccessToken`.

Anyone have any better thoughts on this? I'd rather handle the `DuplicateKeyException` in code...

**Comment by sraza123** [comment]
I had a same issue where duplicate rows were being generated when multiple threads coming into JdbcTokenStore.storeAccessToken at the same time.

I resolved this issue by extending DefaultTokenServices and handle the duplicate key exception.

```
public class CustomTokenServices extends DefaultTokenServices {

    private static final Logger LOGGER = LoggerFactory.getLogger(CustomTokenServices.class);

    @Override
    public OAuth2AccessToken createAccessToken(OAuth2Authentication authentication) throws AuthenticationException {
        try {
            return super.createAccessToken(authentication);
        }catch (DuplicateKeyException dke) {
            LOGGER.info(String.format("Duplicate user found for %s",authentication.getUserAuthentication().getPrincipal()));
            return super.getAccessToken(authentication);
        }catch (Exception ex) {
            LOGGER.info(String.format("Exception while creating access token %s",ex));
        }
        return null;
    }
}
```
Any feedback on this approach is appreciated

**Comment by dcalde** [comment]
Same problem here. I am running a simple load test with just 4 concurrent users trying to authenticate with username/password and getting DuplicateKeyException.

**Comment by dsyer** [comment]
@bilak and I had a long discussion about this on gitter (see the link to his issue above). He found that Postgres required a different workaround than other databases, but a combination of transaction isolation and retry would always work I believe.

Your load test is probably not very realistic though, or if it is, then maybe you should think about how your client apps are implemented (how would one client be sending concurrent requests for the same user?).

**Comment by sraza123** [comment]
@dsyer This is a very straight forward and common case, where the same user is logged in on multiple browser and when they try to hit a request this exception occurs. I have fix it with a way around mentioned above.

**Comment by dsyer** [comment]
OK, but I still don't believe you have the right issue. Please attach a sample to #1033.

**Comment by dcalde** [comment]
I resolved the problem for application using UPSERTS, in my case Postgres 9.5+ INSERT .. ON CONFLICT DO:

```
@Bean
public TokenStore tokenStore() {
    JdbcTokenStore tokenStore = new JdbcTokenStore(dataSource);
    tokenStore.setInsertAccessTokenSql("insert into oauth_access_token as t (token_id, token, authentication_id, user_name, client_id, authentication, refresh_token)\n" +
        "values (?, ?, ?, ?, ?, ?, ?)\n" +
        "on conflict on constraint pk_oauth_access_token  do\n" +
        "update set\n" +
        "    token_id = excluded.token_id,\n" +
        "    token = excluded.token,\n" +
        "    authentication = excluded.authentication,\n" +
        "    refresh_token = excluded.refresh_token\n" +
        "  where t.authentication_id = excluded.authentication_id");
    return tokenStore;
}
```

**Comment by dyhack** [comment]
> I had a same issue where duplicate rows were being generated when multiple threads coming into JdbcTokenStore.storeAccessToken at the same time.
> 
> I resolved this issue by extending DefaultTokenServices and handle the duplicate key exception.
> 
> ```
> public class CustomTokenServices extends DefaultTokenServices {
> 
>     private static final Logger LOGGER = LoggerFactory.getLogger(CustomTokenServices.class);
> 
>     @Override
>     public OAuth2AccessToken createAccessToken(OAuth2Authentication authentication) throws AuthenticationException {
>         try {
>             return super.createAccessToken(authentication);
>         }catch (DuplicateKeyException dke) {
>             LOGGER.info(String.format("Duplicate user found for %s",authentication.getUserAuthentication().getPrincipal()));
>             return super.getAccessToken(authentication);
>         }catch (Exception ex) {
>             LOGGER.info(String.format("Exception while creating access token %s",ex));
>         }
>         return null;
>     }
> }
> ```
> 
> Any feedback on this approach is appreciated


I think this also have a DuplicateKeyException，Because When I get accesstoken ,It will restore into database.
we can use pessimism lock or synchronized  can solve this problem.

`public class LockedTokenServices extends DefaultTokenServices {

	@Override
	public synchronized OAuth2AccessToken createAccessToken(OAuth2Authentication authentication)
			throws AuthenticationException {
		return super.createAccessToken(authentication);
	}
}`

**Comment by jgrandja** [comment]
Closing this as questions are better suited on Stack Overflow. We prefer to use GitHub issues for bugs and enhancements.
